nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,313 @@
+
+        with torch.no_grad():
+            input_spectrograms = self.normalize_spectrograms(input_spectrograms, lengths)
+            target_spectrograms = self.normalize_spectrograms(target_spectrograms, lengths)
+
+        # train discriminator
+        if optimizer_idx == 0:
+            enhanced_spectrograms = self.forward(
+                input_spectrograms=input_spectrograms, lengths=lengths, mixing=True, normalize=False
+            )
+            enhanced_spectrograms = rearrange(enhanced_spectrograms, "b c l -> b 1 c l")
+            fake_logits = self.discriminator(enhanced_spectrograms, input_spectrograms, lengths)
+
+            target_spectrograms_ = rearrange(target_spectrograms, "b c l -> b 1 c l").requires_grad_()
+            real_logits = self.discriminator(target_spectrograms_, input_spectrograms, lengths)
+            d_loss = self.discriminator_loss(real_logits, fake_logits)
+            self.log("d_loss", d_loss, prog_bar=True)
+
+            if batch_idx % self._cfg.gradient_penalty_loss_every_n_steps == 0:
+                gp_loss = self.gradient_penalty_loss(target_spectrograms_, real_logits)
+                self.log("d_loss_gp", gp_loss, prog_bar=True)
+                return d_loss + gp_loss
+
+            return d_loss
+
+        # train generator
+        if optimizer_idx == 1:
+            enhanced_spectrograms = self.forward(
+                input_spectrograms=input_spectrograms, lengths=lengths, mixing=True, normalize=False
+            )
+
+            input_spectrograms = rearrange(input_spectrograms, "b c l -> b 1 c l")
+            enhanced_spectrograms = rearrange(enhanced_spectrograms, "b c l -> b 1 c l")
+
+            fake_logits = self.discriminator(enhanced_spectrograms, input_spectrograms, lengths)
+            g_loss = self.generator_loss(fake_logits)
+            c_loss = self.consistency_loss(input_spectrograms, enhanced_spectrograms, lengths)
+
+            self.log("g_loss", g_loss, prog_bar=True)
+            self.log("c_loss", c_loss, prog_bar=True)
+
+            with torch.no_grad():
+                target_spectrograms = rearrange(target_spectrograms, "b c l -> b 1 c l")
+                self.log_illustration(target_spectrograms, input_spectrograms, enhanced_spectrograms, lengths)
+            return g_loss + c_loss
+
+    def configure_optimizers(self):
+        generator_opt = instantiate(self._cfg.generator_opt, params=self.generator.parameters(),)
+        discriminator_opt = instantiate(self._cfg.discriminator_opt, params=self.discriminator.parameters())
+        return [discriminator_opt, generator_opt], []
+
+    def setup_training_data(self, train_data_config):
+        dataset = instantiate(train_data_config.dataset)
+        self._train_dl = torch.utils.data.DataLoader(
+            dataset, collate_fn=dataset.collate_fn, **train_data_config.dataloader_params
+        )
+
+    def setup_validation_data(self, val_data_config):
```

Comment:
We can go two ways here:
1) Remove `setup_validation_data` because the is no `val_ds` in the config, nor is there a validation step implementation.
OR
2) Add `val_ds` section to config as well as a validation_step to the model class.

Reply:
Let's say we compute discriminator loss on some held-out set. I am extremely skeptical whether it is a sensible way to choose between two models based on that loss. Until proven on a couple of downstream setups which use output of this GAN, of course.
I don't think validation steps are of value here, even though the generation is conditional. A validation step might be introduced in the future, if (when) we get some positive evidence, but for now let's keep it simple.

Replaced with a `pass` and a comment summing up above. This is an `@abstractmethod` in `ModelPT`, so there has to be _some_ implementation

Reply:
Basically went with option 1.